File: core/java/src/main/java/hara/lang/protocol/IContext.java
package hara.lang.protocol;

import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

/** Callable root context with pointer ownership checks. */
@HaraProtocolBinding(namespace = "std.protocol.icontext", name = "IContext")
public interface IContext {
  @HaraMethod(value = "call", arity = -1, variadic = true)
  Object call(Object... args);

  @HaraMethod(value = "has-ptr?", arity = 2)
  default boolean hasPtr(IPointer pointer) {
    return false;
  }
}

File: core/rust/src/lang/protocol/icontext.rs
use hara_protocol_macros::hara_protocol;

#[hara_protocol(namespace = "std.protocol.icontext", name = "IContext")]
pub trait IContext<A, P> {
  type Output;

  #[hara_method(value = "call", arity = -1, variadic = true, min_arity = 1)]
  fn call(&mut self, arguments: A) -> Self::Output;

  #[hara_method(value = "has-ptr?", arity = 2)]
  fn has_ptr(&self, pointer: &P) -> bool;
}
